From ab3de277aed1ac5b3b587b0bb2ad9e614900fbc7 Mon Sep 17 00:00:00 2001
From: Jason Wessel <jason.wessel@windriver.com>
Date: Wed, 10 Sep 2008 11:38:24 -0500
Subject: [PATCH] kgdb: debugger reboot notifier hook

Notify debugger as well as catch reboot events when a kgdb I/O driver
is registered.

Signed-off-by: Jason Wessel <jason.wessel@windriver.com>
---
 kernel/kgdb.c |   39 +++++++++++++++++++++++++++++++++++++++
 1 files changed, 39 insertions(+), 0 deletions(-)

diff --git a/kernel/kgdb.c b/kernel/kgdb.c
index 949806a..42bf8a3 100644
--- a/kernel/kgdb.c
+++ b/kernel/kgdb.c
@@ -1585,11 +1585,49 @@ static struct sysrq_key_op sysrq_gdb_op = {
 };
 #endif
 
+static int
+kgdb_notify_reboot(struct notifier_block *this, unsigned long code, void *x)
+{
+	unsigned long flags;
+
+	if (!kgdb_connected)
+		return 0;
+
+	if (code == SYS_RESTART || code == SYS_HALT || code == SYS_POWER_OFF) {
+		local_irq_save(flags);
+		if (kgdb_io_ops->write_char) {
+			/* Do not use put_packet to avoid hanging
+			 * in case the attached debugger disappeared
+			 * or does not respond timely.
+			 */
+			kgdb_io_ops->write_char('$');
+			kgdb_io_ops->write_char('X');
+			kgdb_io_ops->write_char('0');
+			kgdb_io_ops->write_char('0');
+			kgdb_io_ops->write_char('#');
+			kgdb_io_ops->write_char('b');
+			kgdb_io_ops->write_char('8');
+			if (kgdb_io_ops->flush)
+				kgdb_io_ops->flush();
+		}
+		kgdb_connected = 0;
+		local_irq_restore(flags);
+	}
+	return NOTIFY_DONE;
+}
+
+static struct notifier_block kgdb_reboot_notifier = {
+	.notifier_call		= kgdb_notify_reboot,
+	.next			= NULL,
+	.priority		= INT_MAX,
+};
+
 static void kgdb_register_callbacks(void)
 {
 	if (!kgdb_io_module_registered) {
 		kgdb_io_module_registered = 1;
 		kgdb_arch_init();
+		register_reboot_notifier(&kgdb_reboot_notifier);
 #ifdef CONFIG_MAGIC_SYSRQ
 		register_sysrq_key('g', &sysrq_gdb_op);
 #endif
@@ -1608,6 +1646,7 @@ static void kgdb_unregister_callbacks(void)
 	 * break exceptions at the time.
 	 */
 	if (kgdb_io_module_registered) {
+		unregister_reboot_notifier(&kgdb_reboot_notifier);
 		kgdb_io_module_registered = 0;
 		kgdb_arch_exit();
 #ifdef CONFIG_MAGIC_SYSRQ
-- 
1.5.5.1

